<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Ne estimation</title>


<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div id="section-header">



<h1 class="title toc-ignore">Ne estimation</h1>

</div>


<STYLE type='text/css' scoped>
PRE.fansi SPAN {padding-top: .25em; padding-bottom: .25em};
</STYLE>
<div id="section-lets-get-started" class="section level2">
<h2>Let’s get started</h2>
<p>We need to load some packages here:</p>
<pre class="r"><code>library(slimr)</code></pre>
<pre><code>Welcome to the slimr package for forward population genetics simulation in SLiM. For more information on SLiM please visit https://messerlab.org/slim/ .</code></pre>
<pre><code>
Attaching package: &#39;slimr&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:shiny&#39;:

    code</code></pre>
<pre class="r"><code>library(dartR)</code></pre>
<pre><code>Loading required package: adegenet</code></pre>
<pre><code>Loading required package: ade4</code></pre>
<pre><code>Registered S3 method overwritten by &#39;spdep&#39;:
  method   from
  plot.mst ape </code></pre>
<pre><code>
   /// adegenet 2.1.3 is loaded ////////////

   &gt; overview: &#39;?adegenet&#39;
   &gt; tutorials/doc/questions: &#39;adegenetWeb()&#39; 
   &gt; bug reports/feature requests: adegenetIssues()</code></pre>
<pre><code>Loading required package: ggplot2</code></pre>
<pre><code>Registered S3 method overwritten by &#39;pegas&#39;:
  method      from
  print.amova ade4</code></pre>
<pre><code>Registered S3 method overwritten by &#39;GGally&#39;:
  method from   
  +.gg   ggplot2</code></pre>
<pre><code>Registered S3 method overwritten by &#39;genetics&#39;:
  method      from 
  [.haplotype pegas</code></pre>
<pre class="r"><code>library(vcfR)</code></pre>
<pre><code>
   *****       ***   vcfR   ***       *****
   This is vcfR 1.12.0 
     browseVignettes(&#39;vcfR&#39;) # Documentation
     citation(&#39;vcfR&#39;) # Citation
   *****       *****      *****       *****</code></pre>
<pre class="r"><code>library(parallel)
source(&quot;/cloud/project/bernd_Ne/gl2genepop.r&quot;)
library(tictoc)
library(future)</code></pre>
<pre class="r"><code>## number of snps to simulate
simnsnps &lt;- 5000 </code></pre>
</div>
<div id="section-simulation" class="section level2">
<h2>Simulation</h2>
<p>Simulate 5000 SNPs for a constant ideal population of a set number individuals over 50 generations. We will use set recombination rates (e.g. depending on the number and length of the chromosomes).</p>
</div>
<div id="section-recombination-rate" class="section level2">
<h2>Recombination rate</h2>
<p>The following code is to calculate recombination rate for a set number of snps (simsnps) randomly placed on the GED chromosomes. We take the distance between SNPs (if on the same chromosome) and calculate a recombination rate depending on the distance 1cM per 1 million bases). If on different chromosomes we set the recombination rate between SNPs as “0.5”.</p>
<p>Luckily we have a chromosome map for our species.</p>
<pre class="r"><code>#sample GED chromosom
gedchromo &lt;- read.csv(&quot;/cloud/project/bernd_Ne/GED_chromosomes.csv&quot;) #averaged length of sex chromosome.


total.chromo &lt;- sum(gedchromo$csize)* 1e6  #length in bases

chromo &lt;- sort(sample(0:(total.chromo-1),simnsnps+1, replace=FALSE))
recrates &lt;- (diff(chromo)/1e6)*0.01 #convert to recombination rate (1 cM=0.01 per 1e6 basis)

chromends &lt;- cumsum(gedchromo$csize)*1e6

for (i in 1:length(chromends))
{
  index &lt;- which(chromo[1:(simnsnps-1)]&lt;chromends[i]  &amp; chromo[2:simnsnps]&gt;chromends[i])
  
  recrates[index]&lt;- 0.5
}

#plot(chromo, rep(1,length(chromo)), pch=15)
#abline(v=chromends)


snpends &lt;- 0:4999</code></pre>
</div>
<div id="section-the-slim-template" class="section level2">
<h2>The slim template</h2>
<p>The next bit is to create a slim template that allows us to run our simulation scripts with several parameters. A template is needed if you want to run several simulations with different settings. slimR allows you to easily incorportate R objects (e.g. vectors and matrices into slim scripts).</p>
<p>The script is simulating the following:</p>
<ol style="list-style-type: decimal">
<li>Create a population of size</li>
</ol>
<pre class="r"><code>#define some constants first

tic()  #setting a timer to record the runtime

## number of generations to simulate
simngen&lt;-50  </code></pre>
<p>Use the settings from df (rep, Ne and RR) and also setup the simulation.</p>
<p>The simulation runs Ne individuals for 50 generations (with 20 generation of a burnin with a very high mutation rate and then 30 generations with a low mutation rate to create our SNP data. Please note this needs to be set a bit longer in a real simulation. Here fairly low values to allow the simulations run fast.</p>
<p>We simulate simnsnps (5000) by randomly initialising a sequence of 5000 nucleotides. Then we only allow mutations from A-&gt;C (C-&gt;A) and G-&gt;T (T-&gt;G), which makes sure there are only two allelels in at a single loci.</p>
<p>If RR is set to 0.99 we use the recombination rate calcuated above, otherwise we use a constante recombination rate between SNPs.</p>
<p>At the end of a simulation create a VCF output.</p>
<pre class="r"><code>slim_script(
  slim_block(initialize(),
             {
               defineConstant(&quot;alpha1&quot;, 0.1);
               defineConstant(&quot;alpha2&quot;, 1e-8);
               
               defineConstant(&quot;rates&quot;, slimr_inline(recrates));
               defineConstant(&quot;ends&quot;, slimr_inline(snpends));
               
               
         
               ## m1 mutation type: neutral
               defineConstant(&quot;L&quot;, !!simnsnps);
               defineConstant(&quot;simlen&quot;, !!simngen);
               
               #how many outputs (simlen=only once at the end set here)
               defineConstant(&quot;every&quot;, simlen);
               
               initializeSLiMOptions(nucleotideBased=T);  #define Nucleotid based simulation
               initializeAncestralNucleotides(randomNucleotides(L)); #random starting status (the same for every individual!!!!)
               
               initializeMutationTypeNuc(&quot;m1&quot;, 0.5, &quot;f&quot;,0.0 ); #neutral mutations (dominance coefficient and distribution of fitness effect)
               m1.mutationStackPolicy = &quot;l&quot;; #only last mutation is kept
               m1.convertToSubstitution=T;  #we are allowed to loose loci (if fixed)
               
               ## g1 genomic element type: uses m1 for all mutations
               defineConstant(&quot;mm&quot;, matrix(c(0.0,alpha1,0.0,0.0, alpha1,0.0,0.0,0.0, 0.0,0.0,0.0,alpha1, 0.0,0.0,alpha1,0.0 ),nrow=4, ncol=4));
               defineConstant(&quot;mm2&quot;, matrix(c(0.0,alpha2,0.0,0.0, alpha2,0.0,0.0,0.0, 0.0,0.0,0.0,alpha2, 0.0,0.0,alpha2,0.0 ),nrow=4, ncol=4));
               initializeGenomicElementType(&quot;g1&quot;, m1, 1.0, mm);
               
               initializeGenomicElement(g1, 0, L-1);  #now set up a genome of 5000 snps
               
               #set recombination rates along the genomice element
               crates = rates
               if (slimr_template(&quot;RR&quot;)!=0.99)  crates = rep(slimr_template(&quot;RR&quot;),5000); #change to constant or use GED version
                 
               initializeRecombinationRate(rates = crates, ends = ends)     
               
               
             }),
  slim_block(1, 
             {
               sim.addSubpop(&quot;p1&quot;,slimr_template(&quot;Ne&quot;));
             }),
  

  
  slim_block(1,simngen,late(), {
                #create an output at the end only here
               if (sim.generation / every== round(sim.generation/every)) {
               pre = &quot;&quot;
               if (sim.generation&lt;100) pre=&quot;0&quot; 
               if (sim.generation&lt;10) pre=&quot;00&quot; 
               nn = paste0( c(&quot;/cloud/project/bernd_Ne/Neest/out&quot;,&quot;_Ne_&quot;,slimr_template(&quot;Ne&quot;),&quot;_rep_&quot;,slimr_template(&quot;rep&quot;),&quot;_RR_&quot;,slimr_template(&quot;RR&quot;),&quot;_&quot;,pre,sim.generation,&quot;.vcf&quot;)) ;
               
               p1.individuals.genomes.outputVCF(nn, simplifyNucleotides=T);
              
             } ##end every
             }),
  #set mutation rate to lower values after 20 generations
  slim_block(20, early(), {
    g1.setMutationMatrix(mm2);
  }),
  
  
  #finish the simultion (to free memory)
  slim_block(simngen, late(), {
               sim.simulationFinished();
             })
) -&gt; script_temp</code></pre>
</div>
<div id="section-create-slim-scripts" class="section level2">
<h2>Create slim scripts</h2>
<p>We can now check the template script (which has three place holders [rep, Ne, RR]). If RR=0.99 the script will use the GED recombination rates calculated above, otherwise it will use a constant recombination rate for all SNPs (0.001=highly lined and 0.5=all on different chromosomes)</p>
<pre class="r"><code>#RR = 0.99 = using GED recombination rates
df &lt;- expand.grid(rep=1:3,  Ne = c(10,20), RR=c(0.001,0.5, 0.99))
df</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["rep"],"name":[1],"type":["int"],"align":["right"]},{"label":["Ne"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["RR"],"name":[3],"type":["dbl"],"align":["right"]}],"data":[{"1":"1","2":"10","3":"0.001"},{"1":"2","2":"10","3":"0.001"},{"1":"3","2":"10","3":"0.001"},{"1":"1","2":"20","3":"0.001"},{"1":"2","2":"20","3":"0.001"},{"1":"3","2":"20","3":"0.001"},{"1":"1","2":"10","3":"0.500"},{"1":"2","2":"10","3":"0.500"},{"1":"3","2":"10","3":"0.500"},{"1":"1","2":"20","3":"0.500"},{"1":"2","2":"20","3":"0.500"},{"1":"3","2":"20","3":"0.500"},{"1":"1","2":"10","3":"0.990"},{"1":"2","2":"10","3":"0.990"},{"1":"3","2":"10","3":"0.990"},{"1":"1","2":"20","3":"0.990"},{"1":"2","2":"20","3":"0.990"},{"1":"3","2":"20","3":"0.990"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>We use the function slim_script_render to create 18 template scripts (in a list), that includes now values for the place holders.</p>
<pre class="r"><code>script_temp_df &lt;- slimr_script_render(script_temp, template =df)


script_temp_df[[1]]</code></pre>
<PRE class="fansi fansi-output"><CODE>&lt;slimr_script[5]&gt;
<span style='background-color: #00BBBB;font-weight: bold;'>block_init:</span><span>initialize() </span><span style='color: #BBBB00;'>{</span><span>
    </span><span style='color: #00BBBB;'>defineConstant</span><span style='color: #0000BB;'>(</span><span style='color: #BBBB00;'>"alpha1"</span><span>, </span><span style='color: #0000BB;'>0.1)</span><span>;
    </span><span style='color: #00BBBB;'>defineConstant</span><span style='color: #0000BB;'>(</span><span style='color: #BBBB00;'>"alpha2"</span><span>, </span><span style='color: #0000BB;'>1e-08)</span><span>;
    </span><span style='color: #00BBBB;'>defineConstant</span><span style='color: #0000BB;'>(</span><span style='color: #BBBB00;'>"rates"</span><span>, </span><span style='color: #00BBBB;'>{</span><span>. </span><span style='color: #00BB00;'>&lt;-</span><span> </span><span style='color: #00BBBB;'>c</span><span style='color: #BBBB00;'>(</span><span style='color: #0000BB;'>0.000326</span><span>, </span><span style='color: #0000BB;'>0.001786</span><span>, </span><span style='color: #0000BB;'>0.002023</span><span>, </span><span style='color: #0000BB;'>0.003998</span><span>, </span><span style='color: #0000BB;'>0.005435</span><span>, ...</span><span style='color: #BBBB00;'>)</span><span style='color: #00BBBB;'>}</span><span style='color: #0000BB;'>)</span><span>;
    </span><span style='color: #00BBBB;'>defineConstant</span><span style='color: #0000BB;'>(</span><span style='color: #BBBB00;'>"ends"</span><span>, </span><span style='color: #00BBBB;'>{</span><span>. </span><span style='color: #00BB00;'>&lt;-</span><span> </span><span style='color: #00BBBB;'>c</span><span style='color: #BBBB00;'>(</span><span style='color: #0000BB;'>0</span><span>, </span><span style='color: #0000BB;'>1</span><span>, </span><span style='color: #0000BB;'>2</span><span>, </span><span style='color: #0000BB;'>3</span><span>, </span><span style='color: #0000BB;'>4</span><span>, </span><span style='color: #0000BB;'>5</span><span>, </span><span style='color: #0000BB;'>6</span><span>, </span><span style='color: #0000BB;'>7</span><span>, </span><span style='color: #0000BB;'>8</span><span>, </span><span style='color: #0000BB;'>9</span><span>, ...</span><span style='color: #BBBB00;'>)</span><span style='color: #00BBBB;'>}</span><span style='color: #0000BB;'>)</span><span>;
    </span><span style='color: #00BBBB;'>defineConstant</span><span style='color: #0000BB;'>(</span><span style='color: #BBBB00;'>"L"</span><span>, </span><span style='color: #0000BB;'>5000)</span><span>;
    </span><span style='color: #00BBBB;'>defineConstant</span><span style='color: #0000BB;'>(</span><span style='color: #BBBB00;'>"simlen"</span><span>, </span><span style='color: #0000BB;'>50)</span><span>;
    </span><span style='color: #00BBBB;'>defineConstant</span><span style='color: #0000BB;'>(</span><span style='color: #BBBB00;'>"every"</span><span>, simlen</span><span style='color: #0000BB;'>)</span><span>;
    </span><span style='color: #00BBBB;'>initializeSLiMOptions</span><span style='color: #0000BB;'>(</span><span>nucleotideBased = T</span><span style='color: #0000BB;'>)</span><span>;
    </span><span style='color: #00BBBB;'>initializeAncestralNucleotides</span><span style='color: #0000BB;'>(</span><span style='color: #00BBBB;'>randomNucleotides(</span><span>L</span><span style='color: #00BBBB;'>)</span><span style='color: #0000BB;'>)</span><span>;
    </span><span style='color: #00BBBB;'>initializeMutationTypeNuc</span><span style='color: #0000BB;'>(</span><span style='color: #BBBB00;'>"m1"</span><span>, </span><span style='color: #0000BB;'>0.5</span><span>, </span><span style='color: #BBBB00;'>"f"</span><span>, </span><span style='color: #0000BB;'>0)</span><span>;
    m1.mutationStackPolicy </span><span style='color: #00BB00;'>=</span><span> </span><span style='color: #BBBB00;'>"l"</span><span>;
    m1.convertToSubstitution </span><span style='color: #00BB00;'>=</span><span> T;
    </span><span style='color: #00BBBB;'>defineConstant</span><span style='color: #0000BB;'>(</span><span style='color: #BBBB00;'>"mm"</span><span>, </span><span style='color: #00BBBB;'>matrix(c</span><span style='color: #BBBB00;'>(</span><span style='color: #0000BB;'>0</span><span>, alpha1, </span><span style='color: #0000BB;'>0</span><span>, </span><span style='color: #0000BB;'>0</span><span>, alpha1, </span><span style='color: #0000BB;'>0</span><span>, </span><span style='color: #0000BB;'>0</span><span>, </span><span style='color: #0000BB;'>0</span><span>, </span><span style='color: #0000BB;'>0</span><span>, </span><span style='color: #0000BB;'>0</span><span>, </span><span style='color: #0000BB;'>0</span><span>, alpha1, </span><span style='color: #0000BB;'>0</span><span>, </span><span style='color: #0000BB;'>0</span><span>, alpha1, </span><span style='color: #0000BB;'>0</span><span style='color: #BBBB00;'>)</span><span>, nrow = </span><span style='color: #0000BB;'>4</span><span>, ncol = </span><span style='color: #0000BB;'>4</span><span style='color: #00BBBB;'>)</span><span style='color: #0000BB;'>)</span><span>;
    </span><span style='color: #00BBBB;'>defineConstant</span><span style='color: #0000BB;'>(</span><span style='color: #BBBB00;'>"mm2"</span><span>, </span><span style='color: #00BBBB;'>matrix(c</span><span style='color: #BBBB00;'>(</span><span style='color: #0000BB;'>0</span><span>, alpha2, </span><span style='color: #0000BB;'>0</span><span>, </span><span style='color: #0000BB;'>0</span><span>, alpha2, </span><span style='color: #0000BB;'>0</span><span>, </span><span style='color: #0000BB;'>0</span><span>, </span><span style='color: #0000BB;'>0</span><span>, </span><span style='color: #0000BB;'>0</span><span>, </span><span style='color: #0000BB;'>0</span><span>, </span><span style='color: #0000BB;'>0</span><span>, alpha2, </span><span style='color: #0000BB;'>0</span><span>, </span><span style='color: #0000BB;'>0</span><span>, alpha2, </span><span style='color: #0000BB;'>0</span><span style='color: #BBBB00;'>)</span><span>, nrow = </span><span style='color: #0000BB;'>4</span><span>, ncol = </span><span style='color: #0000BB;'>4</span><span style='color: #00BBBB;'>)</span><span style='color: #0000BB;'>)</span><span>;
    </span><span style='color: #00BBBB;'>initializeGenomicElementType</span><span style='color: #0000BB;'>(</span><span style='color: #BBBB00;'>"g1"</span><span>, m1, </span><span style='color: #0000BB;'>1</span><span>, mm</span><span style='color: #0000BB;'>)</span><span>;
    </span><span style='color: #00BBBB;'>initializeGenomicElement</span><span style='color: #0000BB;'>(</span><span>g1, </span><span style='color: #0000BB;'>0</span><span>, L </span><span style='color: #00BB00;'>-</span><span> </span><span style='color: #0000BB;'>1)</span><span>;
    crates </span><span style='color: #00BB00;'>=</span><span> rates;
    </span><span style='color: #BB0000;'>if</span><span> </span><span style='color: #0000BB;'>(0.001</span><span> != </span><span style='color: #0000BB;'>0.99)</span><span> crates </span><span style='color: #00BB00;'>=</span><span> </span><span style='color: #00BBBB;'>rep</span><span style='color: #0000BB;'>(0.001</span><span>, </span><span style='color: #0000BB;'>5000)</span><span>;
    </span><span style='color: #00BBBB;'>initializeRecombinationRate</span><span style='color: #0000BB;'>(</span><span>rates = crates, ends = ends</span><span style='color: #0000BB;'>)</span><span>;
</span><span style='color: #BBBB00;'>}</span><span>

</span><span style='background-color: #00BBBB;font-weight: bold;'>block_2:</span><span>1 early() </span><span style='color: #BBBB00;'>{</span><span>
    </span><span style='color: #00BBBB;'>sim.addSubpop</span><span style='color: #0000BB;'>(</span><span style='color: #BBBB00;'>"p1"</span><span>, </span><span style='color: #0000BB;'>10)</span><span>;
</span><span style='color: #BBBB00;'>}</span><span>

</span><span style='background-color: #00BBBB;font-weight: bold;'>block_3:</span><span>1:50 late() </span><span style='color: #BBBB00;'>{</span><span>
    </span><span style='color: #BB0000;'>if</span><span> </span><span style='color: #0000BB;'>(</span><span>sim.generation </span><span style='color: #00BB00;'>/</span><span> every </span><span style='color: #00BB00;'>==</span><span> </span><span style='color: #00BBBB;'>round(</span><span>sim.generation </span><span style='color: #00BB00;'>/</span><span> every</span><span style='color: #00BBBB;'>)</span><span style='color: #0000BB;'>)</span><span> </span><span style='color: #0000BB;'>{</span><span>
      pre </span><span style='color: #00BB00;'>=</span><span> </span><span style='color: #BBBB00;'>""</span><span>;
      </span><span style='color: #BB0000;'>if</span><span> </span><span style='color: #00BBBB;'>(</span><span>sim.generation </span><span style='color: #00BB00;'>&lt;</span><span> </span><span style='color: #0000BB;'>100</span><span style='color: #00BBBB;'>)</span><span> pre </span><span style='color: #00BB00;'>=</span><span> </span><span style='color: #BBBB00;'>"0"</span><span>;
      </span><span style='color: #BB0000;'>if</span><span> </span><span style='color: #00BBBB;'>(</span><span>sim.generation </span><span style='color: #00BB00;'>&lt;</span><span> </span><span style='color: #0000BB;'>10</span><span style='color: #00BBBB;'>)</span><span> pre </span><span style='color: #00BB00;'>=</span><span> </span><span style='color: #BBBB00;'>"00"</span><span>;
      nn </span><span style='color: #00BB00;'>=</span><span> </span><span style='color: #00BBBB;'>paste0(c</span><span style='color: #BBBB00;'>("/cloud/project/bernd_Ne/Neest/out"</span><span>, </span><span style='color: #BBBB00;'>"_Ne_"</span><span>, </span><span style='color: #0000BB;'>10</span><span>, </span><span style='color: #BBBB00;'>"_rep_"</span><span>, </span><span style='color: #0000BB;'>1</span><span>, </span><span style='color: #BBBB00;'>"_RR_"</span><span>, </span><span style='color: #0000BB;'>0.001</span><span>, </span><span style='color: #BBBB00;'>"_"</span><span>, pre, sim.generation, </span><span style='color: #BBBB00;'>".vcf")</span><span style='color: #00BBBB;'>)</span><span>;
      </span><span style='color: #00BBBB;'>p1.individuals.genomes.outputVCF(</span><span>nn, simplifyNucleotides = T</span><span style='color: #00BBBB;'>)</span><span>;
    </span><span style='color: #0000BB;'>}</span><span>
</span><span style='color: #BBBB00;'>}</span><span>

</span><span style='background-color: #00BBBB;font-weight: bold;'>block_4:</span><span>20 early() </span><span style='color: #BBBB00;'>{</span><span>
    </span><span style='color: #00BBBB;'>g1.setMutationMatrix</span><span style='color: #0000BB;'>(</span><span>mm2</span><span style='color: #0000BB;'>)</span><span>;
</span><span style='color: #BBBB00;'>}</span><span>

</span><span style='background-color: #00BBBB;font-weight: bold;'>block_5:</span><span>50 late() </span><span style='color: #BBBB00;'>{</span><span>
    </span><span style='color: #00BBBB;'>sim.simulationFinished</span><span style='color: #0000BB;'>()</span><span>;
</span><span style='color: #BBBB00;'>}</span><span>
</span></CODE></PRE>
<p>Now we can run the slim scripts in parallel by defining the number of workers. Please do not use more than 2 within rcloud to save computation hours.</p>
<pre class="r"><code>#if you want to run a single script 
#slim_run(script_temp_df[[1]])
#check the simulation files in the Neest folder

plan(multisession(workers = 2))
sr &lt;- slim_run(script_temp_df,parallel = TRUE) #based on future plan</code></pre>
</div>
<div id="section-check-vcf-outputs" class="section level2">
<h2>check vcf outputs</h2>
<p>We want to check the output files. Here a quick and dirty way to read all files in and get the number of loci and heterozygosity.</p>
<pre class="r"><code>files &lt;- list.files(path=&quot;/cloud/project/bernd_Ne/Neest/&quot;,pattern=&quot;out&quot; )


gls &lt;- list()
for (i in 1:length(files)){
  dumvcf&lt;- read.vcfR(file.path(&quot;/cloud/project//bernd_Ne/Neest&quot;,files[i]),verbose=FALSE)
  gls[[i]]&lt;- vcfR2genlight(dumvcf)
  names(gls)[i] &lt;- files[i]
}  

#some tests/output on the simulated files
#number of loci
unlist(lapply(gls,nLoc))</code></pre>
<pre><code>out_Ne_10_rep_1_RR_0.001_050.vcf   out_Ne_10_rep_1_RR_0.5_050.vcf 
                            1420                             1383 
 out_Ne_10_rep_1_RR_0.99_050.vcf out_Ne_10_rep_2_RR_0.001_050.vcf 
                            1553                             1464 
  out_Ne_10_rep_2_RR_0.5_050.vcf  out_Ne_10_rep_2_RR_0.99_050.vcf 
                            1939                             1688 
out_Ne_10_rep_3_RR_0.001_050.vcf   out_Ne_10_rep_3_RR_0.5_050.vcf 
                            2143                             1782 
 out_Ne_10_rep_3_RR_0.99_050.vcf out_Ne_20_rep_1_RR_0.001_050.vcf 
                            1842                             4020 
  out_Ne_20_rep_1_RR_0.5_050.vcf  out_Ne_20_rep_1_RR_0.99_050.vcf 
                            3925                             3616 
out_Ne_20_rep_2_RR_0.001_050.vcf   out_Ne_20_rep_2_RR_0.5_050.vcf 
                            3844                             3881 
 out_Ne_20_rep_2_RR_0.99_050.vcf out_Ne_20_rep_3_RR_0.001_050.vcf 
                            3680                             3850 
  out_Ne_20_rep_3_RR_0.5_050.vcf  out_Ne_20_rep_3_RR_0.99_050.vcf 
                            3684                             3670 </code></pre>
<pre class="r"><code>#number of individuals
unlist(lapply(gls,nInd))</code></pre>
<pre><code>out_Ne_10_rep_1_RR_0.001_050.vcf   out_Ne_10_rep_1_RR_0.5_050.vcf 
                              10                               10 
 out_Ne_10_rep_1_RR_0.99_050.vcf out_Ne_10_rep_2_RR_0.001_050.vcf 
                              10                               10 
  out_Ne_10_rep_2_RR_0.5_050.vcf  out_Ne_10_rep_2_RR_0.99_050.vcf 
                              10                               10 
out_Ne_10_rep_3_RR_0.001_050.vcf   out_Ne_10_rep_3_RR_0.5_050.vcf 
                              10                               10 
 out_Ne_10_rep_3_RR_0.99_050.vcf out_Ne_20_rep_1_RR_0.001_050.vcf 
                              10                               20 
  out_Ne_20_rep_1_RR_0.5_050.vcf  out_Ne_20_rep_1_RR_0.99_050.vcf 
                              20                               20 
out_Ne_20_rep_2_RR_0.001_050.vcf   out_Ne_20_rep_2_RR_0.5_050.vcf 
                              20                               20 
 out_Ne_20_rep_2_RR_0.99_050.vcf out_Ne_20_rep_3_RR_0.001_050.vcf 
                              20                               20 
  out_Ne_20_rep_3_RR_0.5_050.vcf  out_Ne_20_rep_3_RR_0.99_050.vcf 
                              20                               20 </code></pre>
<pre class="r"><code>#mean observed Heterozygosity for all 18 simulations
plot(1:length(gls),unlist(lapply(gls, function(x) mean(gl.Ho(x)))),pch=16)</code></pre>
<p><img src="Ne_estimation_files/figure-html/unnamed-chunk-8-1.png" width="80%" /></p>
</div>
<div id="section-run-neestimator-from-within-r" class="section level2">
<h2>Run Neestimator from within R</h2>
<p>Now that we have our 18 simulations the next step is the following:</p>
<ol style="list-style-type: decimal">
<li>sample different levels of SNPs (snplevels)</li>
<li>sample different levels of individuals (sslevels)</li>
<li>Run Neestimator over all levels for all 18 simulations</li>
<li>Collect the results from the Nestimator results into a data.frame</li>
<li>Visualise the results.</li>
</ol>
<p>The idea here is the following. steps 1-3 are wrapped into a simulation function. Then we run the simulation function via an mcapply, which allows to run the jobs in parallel if needed.</p>
<pre class="r"><code>snplevels &lt;- c(100,200)  #snp levels

sslevels &lt;- c(5,10)  #numberof individuals

##the simulation function
sim &lt;- function(i)
{
  dumvcf&lt;- read.vcfR(file.path(&quot;/cloud/project//bernd_Ne/Neest/&quot;,files[i]),verbose=FALSE)
  gls&lt;- vcfR2genlight(dumvcf)
  namesgls &lt;- files[i]  
  
  
res &lt;- data.frame(rep=NA, Nesim=NA, nGen=NA, nSnp=NA, sample=NA,RR=NA, Nee0=NA, Nee0_low=NA, Nee0_up=NA )
cc &lt;-1
  

Nesim=as.numeric(gsub(&quot;out_Ne_([0-9]+).*$&quot;, &quot;\\1&quot;, namesgls))

nrep &lt;-as.numeric(gsub(&quot;.*?rep_([0-9]+).*$&quot;, &quot;\\1&quot;, namesgls))
nGen &lt;-  as.numeric(gsub(&quot;.*_([0-9]+).vcf$&quot;, &quot;\\1&quot;, namesgls))
nRR &lt;- as.numeric(gsub(&quot;.*?RR_(0.[0-9]+).*$&quot;, &quot;\\1&quot;, namesgls))

glout &lt;- gls
glout &lt;- gl.filter.monomorphs(glout, verbose = 0)
pop(glout)&lt;- rep(&quot;A&quot;, nInd(glout))

for (ii in 1:length(snplevels))
{
 
  nSnp=snplevels[ii] 
  if(nSnp&lt;=nLoc(glout)) {
for (iii in 1:length(sslevels))
{
  
  ss = sslevels[iii]
  if (ss&lt;=nInd(glout))
  {

#subsample the simulated population
gldummy &lt;- glout[sample(1:nInd(glout), ss,replace = FALSE),sample(1:nLoc(glout), nSnp, replace = FALSE)]


#create a genepop files
gl2genepop(gldummy,paste0(&quot;/cloud/project/bernd_Ne/Neest/slimne&quot;,i,&quot;.gen&quot;))

#adjust the info file for Neeest
inf &lt;- readLines(&quot;/cloud/project/bernd_Ne/Neest/info&quot;)
inf[3]&lt;-paste0(&quot;slimne&quot;,i,&quot;.gen&quot;)
inf[6]&lt;- paste0(&quot;slimLD&quot;,i,&quot;.txt&quot;)
con &lt;- file(paste0(&quot;./Neest/info&quot;,i),&quot;w&quot;)
writeLines(inf, con)
close(con)

#run the Neestimator
system(paste0(&quot;/cloud/project/bernd_Ne/Neest/Ne2-1L i:/cloud/project/bernd_Ne/Neest/info&quot;,i))

ll &lt;- readLines(paste0(&quot;/cloud/project/bernd_Ne/Neest/slimLD&quot;,i,&quot;.txt&quot;))
index &lt;- grep(&quot;Estimated Ne&quot;, ll)
nes &lt;- unlist(strsplit(ll[index],split = &quot;[ ]+&quot;))
index &lt;- grep(&quot;JackKnife on Samples&quot;, ll)
neslow &lt;- unlist(strsplit(ll[index],split = &quot;[ ]+&quot;))
nesup  &lt;- unlist(strsplit(ll[index+1],split = &quot;[ ]+&quot;))
Nee0 &lt;- as.numeric(nes[length(nes)])
Nee0_low &lt;- as.numeric(neslow[length(neslow)])
Nee0_up &lt;- as.numeric(nesup[length(nesup)])

res[cc,]&lt;- c(nrep, Nesim, nGen, nSnp, ss, nRR, Nee0, Nee0_low, Nee0_up)
cc &lt;- cc+1
} #endif ss&lt;= nInd
} #end iii
  } #if snplevels &lt; nLoc
} #end ii (nSnps)
return(res)
} #end sim function

system.time(res &lt;- mclapply(1:length(files), function(x) sim(x), mc.cores = 1))</code></pre>
<pre><code>Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion

Warning in sim(x): NAs introduced by coercion</code></pre>
<pre><code>   user  system elapsed 
  9.312   0.162   9.478 </code></pre>
</div>
<div id="section-collect-the-results" class="section level2">
<h2>Collect the results</h2>
<p>All the is needed todo is now collect the results from the data.frame and plot</p>
<p>Step 4 and 5 below:</p>
<pre class="r"><code>#unlist the output from the parallel runs
res&lt;- do.call(&quot;rbind&quot;,res)

#in case you wan to to save the results
#write.csv(res,&quot;res_sim18.csv&quot;)



ggplot(res, aes(x=factor(Nesim), y=Nee0/Nesim,fill=factor(sample)))+geom_boxplot()+facet_grid(RR~nSnp)+ylim(c(0,1))</code></pre>
<pre><code>Warning: Removed 30 rows containing non-finite values (stat_boxplot).</code></pre>
<p><img src="Ne_estimation_files/figure-html/unnamed-chunk-10-1.png" width="80%" /></p>
<pre class="r"><code>#stop the clock
toc()</code></pre>
<pre><code>34.021 sec elapsed</code></pre>

<script type="application/shiny-prerendered" data-context="server-start">
library(learnr)
knitr::opts_chunk$set(echo = TRUE, comment = "",
                      error = TRUE,
                      fig.width = 7,
                      fig.height = 6,
                      out.width = "80%")
</script>
 
<script type="application/shiny-prerendered" data-context="server">
learnr:::register_http_handlers(session, metadata = NULL)
</script>
 
<script type="application/shiny-prerendered" data-context="server">
session$onSessionEnded(function() {
        learnr:::session_stop_event(session)
      })
</script>
 <!--html_preserve-->
<script type="application/shiny-prerendered" data-context="dependencies">
{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["header-attrs"]},{"type":"character","attributes":{},"value":["2.7"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/pandoc"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["header-attrs.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.7"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["jquery"]},{"type":"character","attributes":{},"value":["1.11.3"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/jquery"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquery.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.7"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["bootstrap"]},{"type":"character","attributes":{},"value":["3.3.5"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/bootstrap"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["viewport"]}},"value":[{"type":"character","attributes":{},"value":["width=device-width, initial-scale=1"]}]},{"type":"character","attributes":{},"value":["js/bootstrap.min.js","shim/html5shiv.min.js","shim/respond.min.js"]},{"type":"character","attributes":{},"value":["css/bootstrap.min.css"]},{"type":"character","attributes":{},"value":["<style>h1 {font-size: 34px;}\n       h1.title {font-size: 38px;}\n       h2 {font-size: 30px;}\n       h3 {font-size: 24px;}\n       h4 {font-size: 18px;}\n       h5 {font-size: 16px;}\n       h6 {font-size: 12px;}\n       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}\n       pre:not([class]) { background-color: white }<\/style>"]},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.7"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["jquery"]},{"type":"character","attributes":{},"value":["1.11.3"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/jquery"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquery.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.7"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["navigation"]},{"type":"character","attributes":{},"value":["1.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/navigation-1.1"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tabsets.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.7"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["highlightjs"]},{"type":"character","attributes":{},"value":["9.12.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/highlightjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["highlight.js"]},{"type":"character","attributes":{},"value":["default.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.7"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["pagedtable"]},{"type":"character","attributes":{},"value":["1.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/pagedtable-1.1"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["js/pagedtable.js"]},{"type":"character","attributes":{},"value":["css/pagedtable.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.7"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["jquery"]},{"type":"character","attributes":{},"value":["1.11.3"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/jquery"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquery.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.7"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["font-awesome"]},{"type":"character","attributes":{},"value":["5.1.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/fontawesome"]}]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["css/all.css","css/v4-shims.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.7"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["bootbox"]},{"type":"character","attributes":{},"value":["4.4.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/bootbox"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["bootbox.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["idb-keyvalue"]},{"type":"character","attributes":{},"value":["3.2.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/idb-keyval"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["idb-keyval-iife-compat.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial"]},{"type":"character","attributes":{},"value":["0.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial.js"]},{"type":"character","attributes":{},"value":["tutorial.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-autocompletion"]},{"type":"character","attributes":{},"value":["0.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-autocompletion.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-diagnostics"]},{"type":"character","attributes":{},"value":["0.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-diagnostics.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]}]}
</script>
<!--/html_preserve-->
<!--html_preserve-->
<script type="application/shiny-prerendered" data-context="execution_dependencies">
{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["packages"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["packages","version"]},"class":{"type":"character","attributes":{},"value":["data.frame"]},"row.names":{"type":"integer","attributes":{},"value":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149]}},"value":[{"type":"character","attributes":{},"value":["ade4","adegenet","ape","base","boot","bslib","calibrate","class","classInt","cluster","coda","codetools","colorspace","combinat","compiler","crayon","dartR","data.table","datasets","DBI","deldir","DEoptimR","digest","dismo","doParallel","dplyr","e1071","ellipsis","evaluate","expm","fansi","farver","fastmap","foreach","furrr","future","gap","gdata","gdistance","gdsfmt","generics","genetics","GGally","ggplot2","globals","glue","gmodels","graphics","grDevices","grid","gridExtra","gtable","gtools","hierfstat","highr","hms","htmltools","htmlwidgets","httpuv","igraph","iterators","jquerylib","jsonlite","KernSmooth","knitr","labeling","later","lattice","LearnBayes","learnr","lifecycle","listenv","magrittr","markdown","MASS","Matrix","memuse","methods","mgcv","mime","mmod","munsell","mvtnorm","nlme","parallel","parallelly","pegas","permute","pillar","pinfsc50","pkgconfig","plyr","png","PopGenReport","prettycode","prettyunits","processx","progress","promises","ps","purrr","R.methodsS3","R.oo","R.utils","R6","raster","RColorBrewer","Rcpp","reshape","reshape2","rgdal","RgoogleMaps","rlang","rmarkdown","robustbase","rprojroot","rstudioapi","sass","scales","seqinr","sf","shiny","slimr","SNPRelate","sp","spData","spdep","splines","StAMPP","stats","stringi","stringr","tibble","tictoc","tidyr","tidyselect","tools","units","utf8","utils","vcfR","vctrs","vegan","viridisLite","withr","xfun","xtable","yaml","zeallot"]},{"type":"character","attributes":{},"value":["1.7-16","2.1.3","5.4-1","4.0.3","1.3-25","0.2.4","1.7.7","7.3-17","0.4-3","2.1.0","0.19-4","0.2-16","2.0-0","0.0-8","4.0.3","1.4.1","1.8.3","1.14.0","4.0.3","1.1.1","0.2-10","1.0-8","0.6.27","1.3-3","1.0.16","1.0.4","1.7-4","0.3.1","0.14","0.999-6","0.4.2","2.1.0","1.1.0","1.5.1","0.2.2","1.21.0","1.2.2","2.18.0","1.3-6","1.26.1","0.1.0","1.3.8.1.3","2.1.0","3.3.3","0.14.0","1.4.2","2.18.1","4.0.3","4.0.3","4.0.3","2.3","0.3.0","3.8.2","0.5-7","0.8","1.0.0","0.5.1.1","1.5.3","1.5.5","1.2.6","1.0.13","0.1.3","1.7.2","2.23-17","1.31","0.4.2","1.1.0.1","0.20-41","2.15.1","0.10.1","1.0.0","0.8.0","2.0.1","1.1","7.3-53","1.2-18","4.1-0","4.0.3","1.8-33","0.10","1.3.3","0.5.0","1.1-1","3.1-149","4.0.3","1.23.0","0.14","0.9-5","1.5.0","1.2.0","2.0.3","1.8.6","0.1-7","3.0.4","1.1.0","1.1.1","3.4.5","1.2.2","1.2.0.1","1.6.0","0.3.4","1.8.1","1.24.0","2.10.1","2.5.0","3.4-5","1.1-2","1.0.6","0.8.8","1.4.4","1.5-23","1.4.5.3","0.4.10","2.7","0.93-7","2.0.2","0.13","0.3.1","1.1.1","4.2-5","0.9-7","1.6.0","0.0.0.9001","1.24.0","1.4-5","0.3.8","1.1-5","4.0.3","1.6.1","4.0.3","1.5.3","1.4.0","3.1.0","1.0","1.1.3","1.1.0","4.0.3","0.7-0","1.1.4","4.0.3","1.12.0","0.3.6","2.5-7","0.3.0","2.4.1","0.21","1.8-4","2.2.1","0.1.0"]}]}]}
</script>
<!--/html_preserve-->
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("section-TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
