---
title: "Ne estimation"
output:
  html_document:
    df_print: paged
  learnr::tutorial: default
  pdf_document: default
runtime: shiny_prerendered
---

```{r setup, include=FALSE}
library(learnr)
knitr::opts_chunk$set(echo = TRUE, comment = "",
                      error = TRUE,
                      fig.width = 7,
                      fig.height = 6,
                      out.width = "80%")
```

```{r color, echo = FALSE, results='asis'}
# crayon needs to be explicitly activated in Rmd
options(crayon.enabled = TRUE)
# Hooks needs to be set to deal with outputs
# thanks to fansi logic
if(requireNamespace("fansi", quietly = TRUE)) {
  old_hooks <- fansi::set_knit_hooks(knitr::knit_hooks, 
                                     which = c("output", "message", "error"))
}
```

## Let's get started

We need to load some packages here:

```{r load_packages}
library(slimr)
library(dartR)
library(vcfR)
library(parallel)
source("/cloud/project/bernd_Ne/gl2genepop.r")
library(tictoc)
library(future)
```

```{r}
## number of snps to simulate
simnsnps <- 5000 
```

## Simulation

Simulate 5000 SNPs for a constant ideal population of a set number individuals over 50 generations. We will use set recombination rates (e.g. depending on the number and length of the chromosomes).

## Recombination rate

The following code is to calculate recombination rate for a set number of snps (simsnps) randomly placed on the GED chromosomes. We take the distance between SNPs (if on the same chromosome) and calculate a recombination rate depending on the distance 1cM per 1 million bases). If on different chromosomes we set the recombination rate between SNPs as "0.5".

Luckily we have a chromosome map for our species.

```{r}
#sample GED chromosom
gedchromo <- read.csv("/cloud/project/bernd_Ne/GED_chromosomes.csv") #averaged length of sex chromosome.


total.chromo <- sum(gedchromo$csize)* 1e6  #length in bases

chromo <- sort(sample(0:(total.chromo-1),simnsnps+1, replace=FALSE))
recrates <- (diff(chromo)/1e6)*0.01 #convert to recombination rate (1 cM=0.01 per 1e6 basis)

chromends <- cumsum(gedchromo$csize)*1e6

for (i in 1:length(chromends))
{
  index <- which(chromo[1:(simnsnps-1)]<chromends[i]  & chromo[2:simnsnps]>chromends[i])
  
  recrates[index]<- 0.5
}

#plot(chromo, rep(1,length(chromo)), pch=15)
#abline(v=chromends)


snpends <- 0:4999


```

## The slim template

The next bit is to create a slim template that allows us to run our simulation scripts with several parameters. A template is needed if you want to run several simulations with different settings. slimR allows you to easily incorportate R objects (e.g. vectors and matrices into slim scripts).

The script is simulating the following:

1.  Create a population of size

```{r}
#define some constants first

tic()  #setting a timer to record the runtime

## number of generations to simulate
simngen<-50  

```

Use the settings from df (rep, Ne and RR) and also setup the simulation.

The simulation runs Ne individuals for 50 generations (with 20 generation of a burnin with a very high mutation rate and then 30 generations with a low mutation rate to create our SNP data. Please note this needs to be set a bit longer in a real simulation. Here fairly low values to allow the simulations run fast.

We simulate simnsnps (5000) by randomly initialising a sequence of 5000 nucleotides. Then we only allow mutations from A-\>C (C-\>A) and G-\>T (T-\>G), which makes sure there are only two allelels in at a single loci.

If RR is set to 0.99 we use the recombination rate calcuated above, otherwise we use a constante recombination rate between SNPs.

At the end of a simulation create a VCF output.

```{r}
slim_script(
  slim_block(initialize(),
             {
               defineConstant("alpha1", 0.1);
               defineConstant("alpha2", 1e-8);
               
               defineConstant("rates", slimr_inline(recrates));
               defineConstant("ends", slimr_inline(snpends));
               
               
         
               ## m1 mutation type: neutral
               defineConstant("L", !!simnsnps);
               defineConstant("simlen", !!simngen);
               
               #how many outputs (simlen=only once at the end set here)
               defineConstant("every", simlen);
               
               initializeSLiMOptions(nucleotideBased=T);  #define Nucleotid based simulation
               initializeAncestralNucleotides(randomNucleotides(L)); #random starting status (the same for every individual!!!!)
               
               initializeMutationTypeNuc("m1", 0.5, "f",0.0 ); #neutral mutations (dominance coefficient and distribution of fitness effect)
               m1.mutationStackPolicy = "l"; #only last mutation is kept
               m1.convertToSubstitution=T;  #we are allowed to loose loci (if fixed)
               
               ## g1 genomic element type: uses m1 for all mutations
               defineConstant("mm", matrix(c(0.0,alpha1,0.0,0.0, alpha1,0.0,0.0,0.0, 0.0,0.0,0.0,alpha1, 0.0,0.0,alpha1,0.0 ),nrow=4, ncol=4));
               defineConstant("mm2", matrix(c(0.0,alpha2,0.0,0.0, alpha2,0.0,0.0,0.0, 0.0,0.0,0.0,alpha2, 0.0,0.0,alpha2,0.0 ),nrow=4, ncol=4));
               initializeGenomicElementType("g1", m1, 1.0, mm);
               
               initializeGenomicElement(g1, 0, L-1);  #now set up a genome of 5000 snps
               
               #set recombination rates along the genomice element
               crates = rates
               if (slimr_template("RR")!=0.99)  crates = rep(slimr_template("RR"),5000); #change to constant or use GED version
                 
               initializeRecombinationRate(rates = crates, ends = ends)     
               
               
             }),
  slim_block(1, 
             {
               sim.addSubpop("p1",slimr_template("Ne"));
             }),
  

  
  slim_block(1,simngen,late(), {
                #create an output at the end only here
               if (sim.generation / every== round(sim.generation/every)) {
               pre = ""
               if (sim.generation<100) pre="0" 
               if (sim.generation<10) pre="00" 
               nn = paste0( c("/cloud/project/bernd_Ne/Neest/out","_Ne_",slimr_template("Ne"),"_rep_",slimr_template("rep"),"_RR_",slimr_template("RR"),"_",pre,sim.generation,".vcf")) ;
               
               p1.individuals.genomes.outputVCF(nn, simplifyNucleotides=T);
              
             } ##end every
             }),
  #set mutation rate to lower values after 20 generations
  slim_block(20, early(), {
    g1.setMutationMatrix(mm2);
  }),
  
  
  #finish the simultion (to free memory)
  slim_block(simngen, late(), {
               sim.simulationFinished();
             })
) -> script_temp

```

## Create slim scripts

We can now check the template script (which has three place holders [rep, Ne, RR]). If RR=0.99 the script will use the GED recombination rates calculated above, otherwise it will use a constant recombination rate for all SNPs (0.001=highly lined and 0.5=all on different chromosomes)

```{r}
#RR = 0.99 = using GED recombination rates
df <- expand.grid(rep=1:3,  Ne = c(10,20), RR=c(0.001,0.5, 0.99))
df

```

We use the function slim_script_render to create 18 template scripts (in a list), that includes now values for the place holders.

```{r}

script_temp_df <- slimr_script_render(script_temp, template =df)


script_temp_df[[1]]
```

Now we can run the slim scripts in parallel by defining the number of workers. Please do not use more than 2 within rcloud to save computation hours.

```{r}

#if you want to run a single script 
#slim_run(script_temp_df[[1]])
#check the simulation files in the Neest folder

plan(multisession(workers = 2))
sr <- slim_run(script_temp_df,parallel = TRUE) #based on future plan


```

## check vcf outputs

We want to check the output files. Here a quick and dirty way to read all files in and get the number of loci and heterozygosity.

```{r}

files <- list.files(path="/cloud/project/bernd_Ne/Neest/",pattern="out" )


gls <- list()
for (i in 1:length(files)){
  dumvcf<- read.vcfR(file.path("/cloud/project//bernd_Ne/Neest",files[i]),verbose=FALSE)
  gls[[i]]<- vcfR2genlight(dumvcf)
  names(gls)[i] <- files[i]
}  

#some tests/output on the simulated files
#number of loci
unlist(lapply(gls,nLoc))

#number of individuals
unlist(lapply(gls,nInd))

#mean observed Heterozygosity for all 18 simulations
plot(1:length(gls),unlist(lapply(gls, function(x) mean(gl.Ho(x)))),pch=16)

```

## Run Neestimator from within R

Now that we have our 18 simulations the next step is the following:

1.  sample different levels of SNPs (snplevels)
2.  sample different levels of individuals (sslevels)
3.  Run Neestimator over all levels for all 18 simulations
4.  Collect the results from the Nestimator results into a data.frame
5.  Visualise the results.

The idea here is the following. steps 1-3 are wrapped into a simulation function. Then we run the simulation function via an mcapply, which allows to run the jobs in parallel if needed.

```{r}



snplevels <- c(100,200)  #snp levels

sslevels <- c(5,10)  #numberof individuals

##the simulation function
sim <- function(i)
{
  dumvcf<- read.vcfR(file.path("/cloud/project//bernd_Ne/Neest/",files[i]),verbose=FALSE)
  gls<- vcfR2genlight(dumvcf)
  namesgls <- files[i]  
  
  
res <- data.frame(rep=NA, Nesim=NA, nGen=NA, nSnp=NA, sample=NA,RR=NA, Nee0=NA, Nee0_low=NA, Nee0_up=NA )
cc <-1
  

Nesim=as.numeric(gsub("out_Ne_([0-9]+).*$", "\\1", namesgls))

nrep <-as.numeric(gsub(".*?rep_([0-9]+).*$", "\\1", namesgls))
nGen <-  as.numeric(gsub(".*_([0-9]+).vcf$", "\\1", namesgls))
nRR <- as.numeric(gsub(".*?RR_(0.[0-9]+).*$", "\\1", namesgls))

glout <- gls
glout <- gl.filter.monomorphs(glout, verbose = 0)
pop(glout)<- rep("A", nInd(glout))

for (ii in 1:length(snplevels))
{
 
  nSnp=snplevels[ii] 
  if(nSnp<=nLoc(glout)) {
for (iii in 1:length(sslevels))
{
  
  ss = sslevels[iii]
  if (ss<=nInd(glout))
  {

#subsample the simulated population
gldummy <- glout[sample(1:nInd(glout), ss,replace = FALSE),sample(1:nLoc(glout), nSnp, replace = FALSE)]


#create a genepop files
gl2genepop(gldummy,paste0("/cloud/project/bernd_Ne/Neest/slimne",i,".gen"))

#adjust the info file for Neeest
inf <- readLines("/cloud/project/bernd_Ne/Neest/info")
inf[3]<-paste0("slimne",i,".gen")
inf[6]<- paste0("slimLD",i,".txt")
con <- file(paste0("./Neest/info",i),"w")
writeLines(inf, con)
close(con)

#run the Neestimator
system(paste0("/cloud/project/bernd_Ne/Neest/Ne2-1L i:/cloud/project/bernd_Ne/Neest/info",i))

ll <- readLines(paste0("/cloud/project/bernd_Ne/Neest/slimLD",i,".txt"))
index <- grep("Estimated Ne", ll)
nes <- unlist(strsplit(ll[index],split = "[ ]+"))
index <- grep("JackKnife on Samples", ll)
neslow <- unlist(strsplit(ll[index],split = "[ ]+"))
nesup  <- unlist(strsplit(ll[index+1],split = "[ ]+"))
Nee0 <- as.numeric(nes[length(nes)])
Nee0_low <- as.numeric(neslow[length(neslow)])
Nee0_up <- as.numeric(nesup[length(nesup)])

res[cc,]<- c(nrep, Nesim, nGen, nSnp, ss, nRR, Nee0, Nee0_low, Nee0_up)
cc <- cc+1
} #endif ss<= nInd
} #end iii
  } #if snplevels < nLoc
} #end ii (nSnps)
return(res)
} #end sim function

system.time(res <- mclapply(1:length(files), function(x) sim(x), mc.cores = 1))

```

## Collect the results

All the is needed todo is now collect the results from the data.frame and plot

Step 4 and 5 below:

```{r}

#unlist the output from the parallel runs
res<- do.call("rbind",res)

#in case you wan to to save the results
#write.csv(res,"res_sim18.csv")



ggplot(res, aes(x=factor(Nesim), y=Nee0/Nesim,fill=factor(sample)))+geom_boxplot()+facet_grid(RR~nSnp)+ylim(c(0,1))


#stop the clock
toc()
```
