---
title: "What Simulations Can Reveal about Linked Selection"
output:
  html_document:
    df_print: paged
  learnr::tutorial: default
  pdf_document: default
runtime: shiny_prerendered
bibliography: references.bib
---

```{r setup, include=FALSE}
library(learnr)
knitr::opts_chunk$set(echo = TRUE, comment = "",
                      error = TRUE,
                      fig.width = 7,
                      fig.height = 6,
                      out.width = "80%")
```

```{r color, include=FALSE, results='asis'}
# crayon needs to be explicitly activated in Rmd
options(crayon.enabled = TRUE)
# Hooks needs to be set to deal with outputs
# thanks to fansi logic
if(requireNamespace("fansi", quietly = TRUE)) {
  old_hooks <- fansi::set_knit_hooks(knitr::knit_hooks, 
                                     which = c("output", "message", "error"))
}
```

## Let's get started

We need to load some packages here:

```{r load_packages}
setwd("/cloud/project/Luis_slimR")
path.folder_sim <- getwd()
library(knitr)
library(ggthemes)
library(snpStats)
library(slimr)
library(dplyr)
library(ggplot2)
library(raster)
library(maptools)
library(zoo)
library(viridis)
library(rgeos)
library(data.table)
library(scales)
library(ggnewscale)
library(stringr)
library(parallel)
library(future)
library(furrr)
library(plotly)
source('hierfstat.R')
source('gHap.R')
source('rotate_matrix_2.R')
source('functions.R')
```

# Unlinked loci assumption

Most population genetic models assume that loci are inherited by offspring as independent units from their parents due to:

-   sexual reproduction

-   random mating

-   independent assortment of chromosomes and

-   recombination

However, loci that are close to each other on the same chromosome are often inherited together in clusters of different loci. This non-independent inheritance of loci will cause a non-random association between their alleles causing a linkage disequilibrium (LD) between loci.

# Linked selection

When LD occurs between two loci, one neutral and one under selection, sometimes neutral loci appear as if they are under selection ("linked selection").

Linked selection can cause the loss of genetic variation at a higher rate than expected by neutral models under two different scenarios:

a)  via selective sweeps, where a beneficial allele will be spread across the population along with the alleles that are linked to it [@smith1974], and;

b)  via background selection (BGS), where a deleterious allele will be eliminated from the population along with the alleles linked to it [@Charlesworth1994].

![](linked%20_selection.jpg){width="658"}

# The slim template

We are going to use the Wright--Fisher model:

-   diploid organisms

-   hermaphroditic individuals

-   constant population size

-   reproduce in non-overlapping generations

We are going to simulate first a large population with high mutation rate and for 900 generations to obtain some genetic variation to work with. Then the population will experience a bottleneck, so we can observe how heterozygosity is lost across generations more clearly.

This time we are going just to simulate deleterious mutations and neutral mutations.

```{r}
#define some constants first

Ne <- 200 # population size of the bottleneck
ini_pop_size <- 1000 # initial population size
gens <- 1000 # number of generations 
gen_bottleneck <- gens-100 # generation when bottleneck occurs
s_del <- 0 # selection coefficient od deleterious mutations
dominance <- 0.25 # dominance coefficient of dleterious mutations
chr_length <- 10000 # number of nucleotides of the chromosomes
mut_rate <- 3e-5 # mutation rate 
cM <- 100 # number of centiMorgans in the chromosome
rec_rate <- (cM/100)/chr_length # recombination rate

every_do <- 10 # output every X generation

sims <- slim_script(
  
  slim_block(initialize(), {
    setSeed(42);
    # define nucleotide based simulation
    initializeSLiMOptions(nucleotideBased=T);
    # random nucleotides
    initializeAncestralNucleotides(randomNucleotides(!!chr_length));
    # neutral mutations
    initializeMutationTypeNuc("m1", 0, "f", 0);
    # deleterious mutations
    initializeMutationTypeNuc("m2", !!dominance, "f", !!s_del);
    # setting neutral and deleterious mutations. they are equally likely to arise
    initializeGenomicElementType("g1", c(m1,m2), c(1,1), mmJukesCantor(slimr_template("mut_rate",3e-5)));
    # chromosomes
    initializeGenomicElement(g1, 0, !!chr_length-1);
    # recombiantion rate
    initializeRecombinationRate(!!rec_rate);
  }),

    slim_block(1, {
    # initial population
    sim%.%SLiMSim$addSubpop("p1", !!ini_pop_size);
  }),
  
  slim_block(!!gen_bottleneck, {
    # population bottleneck
    p1.setSubpopulationSize(!!Ne); 
    # setting off mutations
    sim.chromosome.setHotspotMap(0);
  }),
  
  # OUTPUT EVERY X GENERATION
  slim_block(1,!!gens, late(), {
    # report heterozygosity of loci with neutral mutations
    slimr_output(calcHeterozygosity(p1.genomes,muts=sim.mutationsOfType(m1)), name = "het_neutral", do_every  =!!every_do);
    # report heterozygosity of loci with deleterious mutations
    slimr_output(calcHeterozygosity(p1.genomes,muts=sim.mutationsOfType(m2)), name = "het_del", do_every  =!!every_do);
    # counting number of neutral mutations
    slimr_output(sim.countOfMutationsOfType(m1), name= "neutral", do_every  =!!every_do);
    # counting number of neutral mutations
    slimr_output(sim.countOfMutationsOfType(m2), name= "deleterious", do_every  =!!every_do);
  }),
  
  # OUTPUT AT THE END OF THE SIMULATION
  slim_block(!!gens, late(), {
    # the sequence of the two chromosomes of each individual 
    slimr_output_nucleotides_2();
    # positions os deleterious mutations
    slimr_output(p1.genomes.positionsOfMutationsOfType(m2), name= "pos_del");
    # frequencies of deleterious mutations
    slimr_output(sim.mutationFrequencies(p1, sim.mutationsOfType(m2)),name="freq");
    # finish the simulations
    sim.simulationFinished();
  })
  
)

```

## Testing effective population size (Ne)

![](rate_loss.jpg){width="758"}

```{r}
Ne <- 100 # population size of the bottleneck
ini_pop_size <- 1000 # initial population size
gens <- 1000 # number of generations 
gen_bottleneck <- gens-100 # generation when bottleneck occurs
s_del <- 0 # selection coefficient od deleterious mutations
dominance <- 0.25 # dominance coefficient of dleterious mutations
chr_length <- 10000 # number of nucleotides of the chromosomes
mut_rate <- 3e-5 # mutation rate 
cM <- 50 # number of centiMorgans in the chromosome
rec_rate <- (cM/100)/chr_length # recombination rate

every_do <- 10 # output every X generation
replicates <- 5 # number of simulations replicates

rate_of_loss <- 1 - (1 / (2 * Ne))

sims_test_Ne <- slim_script(
  
  slim_block(initialize(), {
    setSeed(76666);
    # define nucleotide based simulation
    initializeSLiMOptions(nucleotideBased=T);
    # random nucleotides
    initializeAncestralNucleotides(randomNucleotides(!!chr_length));
    # neutral mutations
    initializeMutationTypeNuc("m1", 0, "f", 0);
    # deleterious mutations
    initializeMutationTypeNuc("m2", !!dominance, "f", !!s_del);
    # setting neutral and deleterious mutations. they are equally likely to arise
    initializeGenomicElementType("g1", c(m1,m2), c(1,1), mmJukesCantor(slimr_template("mut_rate",3e-5)));
    # chromosomes
    initializeGenomicElement(g1, 0, !!chr_length-1);
    # recombiantion rate
    initializeRecombinationRate(!!rec_rate);
  }),

    slim_block(1, {
    # initial population
    sim%.%SLiMSim$addSubpop("p1", !!ini_pop_size);
  }),
  
  slim_block(!!gen_bottleneck, {
    # population bottleneck
    p1.setSubpopulationSize(!!Ne); 
    # setting off mutations
    sim.chromosome.setHotspotMap(0);
  }),
  
  # OUTPUT EVERY X GENERATION
  slim_block(1,!!gens, late(), {
    # report heterozygosity of loci with neutral mutations
    slimr_output(calcHeterozygosity(p1.genomes,muts=sim.mutationsOfType(m1)), name = "het_neutral", do_every  =!!every_do);
    # report heterozygosity of loci with deleterious mutations
    slimr_output(calcHeterozygosity(p1.genomes,muts=sim.mutationsOfType(m2)), name = "het_del", do_every  =!!every_do);
    # counting number of neutral mutations
    slimr_output(sim.countOfMutationsOfType(m1), name= "neutral", do_every  =!!every_do);
    # counting number of neutral mutations
    slimr_output(sim.countOfMutationsOfType(m2), name= "deleterious", do_every  =!!every_do);
  }),
  
  # OUTPUT AT THE END OF THE SIMULATION
  slim_block(!!gens, late(), {
    # the sequence of the two chromosomes of each individual 
    slimr_output_nucleotides_2();
    # positions os deleterious mutations
    slimr_output(p1.genomes.positionsOfMutationsOfType(m2), name= "pos_del");
    # frequencies of deleterious mutations
    slimr_output(sim.mutationFrequencies(p1, sim.mutationsOfType(m2)),name="freq");
    # finish the simulations
    sim.simulationFinished();
  })
  
)
# render script
test_Ne <- slimr_script_render(sims_test_Ne,template = list(mut_rate=3e-5),reps = replicates)
# run the slim scripts in parallel by defining the number of workers. Please do not use more than 2 within rcloud to save computation hours.
plan(multisession(workers = 2))
# sun simulations 
test_Ne_res <- slim_run(test_Ne, throw_error = TRUE,parallel = TRUE)
```

### Extracting data and plotting

```{r}

df_rows <- ((gens-gen_bottleneck)/every_do) + 1
df_replicates_he <- as.data.frame(matrix(nrow =df_rows ,ncol = replicates ))
df_replicates_he_exp <- as.data.frame(matrix(nrow =df_rows ,ncol = replicates ))

for (i in 1:length(test_Ne_res)) {
  df_output <- test_Ne_res[[i]]
  df_output <- as.data.frame(cbind(df_output$output_data$generation,df_output$output_data$data,df_output$output_data$name))
  colnames(df_output) <- c("gen","He","name")
  df_output_het_neutral <- df_output[which(df_output$name=="het_neutral"),]
  df_small_pop <- df_output_het_neutral[which(df_output_het_neutral$gen==gen_bottleneck):nrow(df_output_het_neutral),]
  first_He <- as.numeric(df_output_het_neutral[which(df_output_het_neutral$gen==gen_bottleneck)-1,"He"])
  df_small_pop$rate_loss <- first_He * (rate_of_loss^(as.numeric(df_small_pop$gen)-gen_bottleneck) )
  df_replicates_he[,i] <- as.numeric(df_small_pop$He)
  df_replicates_he_exp[,i] <-  df_small_pop$rate_loss
}

df_replicates_he_res <- rowMeans(df_replicates_he)
df_replicates_he_exp_res <- rowMeans(df_replicates_he_exp)

plot_het <- as.data.frame(cbind(as.numeric(df_small_pop$gen), df_replicates_he_res,df_replicates_he_exp_res))
colnames(plot_het) <- c("gen", "He","he_exp")

all_sim <- print(ggplot(plot_het) +
                   geom_line(aes(x=gen,y=He),colour = "red",size=1) +
                   geom_line(aes(x=gen,y=he_exp),colour = "blue",size=1) +
                   theme_bw(base_size = 18) +
                   scale_fill_hc("darkunica")+
                   labs(x="GENERATIONS", y="He", title=NULL)+
                   theme(legend.title=element_blank())+
                   theme(legend.position =  "bottom") +
                   theme(legend.text=element_text(size=14)))
```

# Background selection (BGS)

```{r}
Ne <- 300 # population size of the bottleneck
ini_pop_size <- 1000 # initial population size
gens <- 1000 # number of generations 
gen_bottleneck <- gens-100 # generation when bottleneck occurs
s_del <- -0.01 # selection coefficient of deleterious mutations
dominance <- 0.5 # dominance coefficient of deleterious mutations
chr_length <- 10000 # number of nucleotides of the chromosomes
mut_rate <- 3e-5 # mutation rate 
cM <- 50 # number of centiMorgans in the chromosome
rec_rate <- (cM/100)/chr_length # recombination rate

every_do <- 10 # output every X generation
replicates <- 5 # number of simulations replicates

rate_of_loss <- 1 - (1 / (2 * Ne))

sims_BGS <- slim_script(
  
  slim_block(initialize(), {
    setSeed(8787);
    # define nucleotide based simulation
    initializeSLiMOptions(nucleotideBased=T);
    # random nucleotides
    initializeAncestralNucleotides(randomNucleotides(!!chr_length));
    # neutral mutations
    initializeMutationTypeNuc("m1", 0, "f", 0);
    # deleterious mutations
    initializeMutationTypeNuc("m2", !!dominance, "f", !!s_del);
    # setting neutral and deleterious mutations. they are equally likely to arise
    initializeGenomicElementType("g1", c(m1,m2), c(1,1), mmJukesCantor(slimr_template("mut_rate",3e-5)));
    # chromosomes
    initializeGenomicElement(g1, 0, !!chr_length-1);
    # recombiantion rate
    initializeRecombinationRate(!!rec_rate);
  }),

    slim_block(1, {
    # initial population
    sim%.%SLiMSim$addSubpop("p1", !!ini_pop_size);
  }),
  
  slim_block(!!gen_bottleneck, {
    # population bottleneck
    p1.setSubpopulationSize(!!Ne); 
    # setting off mutations
    sim.chromosome.setHotspotMap(0);
  }),
  
  # OUTPUT EVERY X GENERATION
  slim_block(1,!!gens, late(), {
    # report heterozygosity of loci with neutral mutations
    slimr_output(calcHeterozygosity(p1.genomes,muts=sim.mutationsOfType(m1)), name = "het_neutral", do_every  =!!every_do);
    # report heterozygosity of loci with deleterious mutations
    slimr_output(calcHeterozygosity(p1.genomes,muts=sim.mutationsOfType(m2)), name = "het_del", do_every  =!!every_do);
    # counting number of neutral mutations
    slimr_output(sim.countOfMutationsOfType(m1), name= "neutral", do_every  =!!every_do);
    # counting number of neutral mutations
    slimr_output(sim.countOfMutationsOfType(m2), name= "deleterious", do_every  =!!every_do);
  }),
  
  # OUTPUT AT THE END OF THE SIMULATION
  slim_block(!!gens, late(), {
    # the sequence of the two chromosomes of each individual 
    slimr_output_nucleotides_2();
    # positions os deleterious mutations
    slimr_output(p1.genomes.positionsOfMutationsOfType(m2), name= "pos_del");
    # frequencies of deleterious mutations
    slimr_output(sim.mutationFrequencies(p1, sim.mutationsOfType(m2)),name="freq");
    # finish the simulations
    sim.simulationFinished();
  })
  
)

# render script
BGS <- slimr_script_render(sims_BGS,template = list(mut_rate=3e-5),reps = replicates)
# run the slim scripts in parallel by defining the number of workers. Please do not use more than 2 within rcloud to save computation hours.
plan(multisession(workers = 2))
# sun simulations 
BGS_res <- slim_run(BGS, throw_error = TRUE,parallel = TRUE)

```

### Extracting data and plotting

```{r}

df_rows <- ((gens-gen_bottleneck)/every_do) + 1
df_replicates_he_BGS <- as.data.frame(matrix(nrow =df_rows ,ncol = replicates ))
df_replicates_he_exp_BGS <- as.data.frame(matrix(nrow =df_rows ,ncol = replicates ))

for (i in 1:length(BGS_res)) {
  df_output_BGS <- BGS_res[[i]]
  df_output_BGS <- as.data.frame(cbind(df_output_BGS$output_data$generation,df_output_BGS$output_data$data,df_output_BGS$output_data$name))
  colnames(df_output_BGS) <- c("gen","He","name")
  df_output_het_neutral_BGS <- df_output_BGS[which(df_output_BGS$name=="het_neutral"),]
  df_small_pop_BGS <- df_output_het_neutral_BGS[which(df_output_het_neutral_BGS$gen==gen_bottleneck):nrow(df_output_het_neutral_BGS),]
  first_He_BGS <- as.numeric(df_output_het_neutral_BGS[which(df_output_het_neutral_BGS$gen==gen_bottleneck)-1,"He"])
  df_small_pop_BGS$rate_loss <- first_He_BGS * (rate_of_loss^(as.numeric(df_small_pop_BGS$gen)-gen_bottleneck) )
  df_replicates_he_BGS[,i] <- as.numeric(df_small_pop_BGS$He)
  df_replicates_he_exp_BGS[,i] <-  df_small_pop_BGS$rate_loss
}

df_replicates_he_res_BGS <- rowMeans(df_replicates_he_BGS)
df_replicates_he_exp_res_BGS <- rowMeans(df_replicates_he_exp_BGS)

plot_het_BGS <- as.data.frame(cbind(as.numeric(df_small_pop_BGS$gen), df_replicates_he_res_BGS,df_replicates_he_exp_res_BGS))
colnames(plot_het_BGS) <- c("gen", "He","he_exp")

all_sim_BGS <- print(ggplot(plot_het_BGS) +
                   geom_line(aes(x=gen,y=He),colour = "red",size=1) +
                   geom_line(aes(x=gen,y=he_exp),colour = "blue",size=1) +
                   theme_bw(base_size = 18) +
                   scale_fill_hc("darkunica")+
                   labs(x="GENERATIONS", y="He", title=NULL)+
                   theme(legend.title=element_blank())+
                   theme(legend.position =  "bottom") +
                   theme(legend.text=element_text(size=14)))

```

## What happens in a small population?

```{r}
Ne <- 20 # population size of the bottleneck
ini_pop_size <- 1000 # initial population size
gens <- 1000 # number of generations 
gen_bottleneck <- gens-100 # generation when bottleneck occurs
s_del <- -0.01 # selection coefficient of deleterious mutations
dominance <- 0.1 # dominance coefficient of deleterious mutations
chr_length <- 10000 # number of nucleotides of the chromosomes
mut_rate <- 3e-5 # mutation rate 
cM <- 10 # number of centiMorgans in the chromosome
rec_rate <- (cM/100)/chr_length # recombination rate

every_do <- 10 # output every X generation
replicates <- 5 # number of simulations replicates

rate_of_loss <- 1 - (1 / (2 * Ne))

sims_AOD <- slim_script(
  
  slim_block(initialize(), {
    setSeed(797);
    # define nucleotide based simulation
    initializeSLiMOptions(nucleotideBased=T);
    # random nucleotides
    initializeAncestralNucleotides(randomNucleotides(!!chr_length));
    # neutral mutations
    initializeMutationTypeNuc("m1", 0, "f", 0);
    # deleterious mutations
    initializeMutationTypeNuc("m2", !!dominance, "f", !!s_del);
    # setting neutral and deleterious mutations. they are equally likely to arise
    initializeGenomicElementType("g1", c(m1,m2), c(1,1), mmJukesCantor(slimr_template("mut_rate",3e-5)));
    # chromosomes
    initializeGenomicElement(g1, 0, !!chr_length-1);
    # recombiantion rate
    initializeRecombinationRate(!!rec_rate);
  }),

    slim_block(1, {
    # initial population
    sim%.%SLiMSim$addSubpop("p1", !!ini_pop_size);
  }),
  
  slim_block(!!gen_bottleneck, {
    # population bottleneck
    p1.setSubpopulationSize(!!Ne); 
    # setting off mutations
    sim.chromosome.setHotspotMap(0);
  }),
  
  # OUTPUT EVERY X GENERATION
  slim_block(1,!!gens, late(), {
    # report heterozygosity of loci with neutral mutations
    slimr_output(calcHeterozygosity(p1.genomes,muts=sim.mutationsOfType(m1)), name = "het_neutral", do_every  =!!every_do);
    # report heterozygosity of loci with deleterious mutations
    slimr_output(calcHeterozygosity(p1.genomes,muts=sim.mutationsOfType(m2)), name = "het_del", do_every  =!!every_do);
    # counting number of neutral mutations
    slimr_output(sim.countOfMutationsOfType(m1), name= "neutral", do_every  =!!every_do);
    # counting number of neutral mutations
    slimr_output(sim.countOfMutationsOfType(m2), name= "deleterious", do_every  =!!every_do);
  }),
  
  # OUTPUT AT THE END OF THE SIMULATION
  slim_block(!!gens, late(), {
    # the sequence of the two chromosomes of each individual 
    slimr_output_nucleotides_2();
    # positions os deleterious mutations
    slimr_output(p1.genomes.positionsOfMutationsOfType(m2), name= "pos_del");
    # frequencies of deleterious mutations
    slimr_output(sim.mutationFrequencies(p1, sim.mutationsOfType(m2)),name="freq");
    # finish the simulations
    sim.simulationFinished();
  })
  
)

# render script
AOD <- slimr_script_render(sims_AOD,template = list(mut_rate=mut_rate),reps = replicates)
# run the slim scripts in parallel by defining the number of workers. Please do not use more than 2 within rcloud to save computation hours.
plan(multisession(workers = 2))
# sun simulations 
AOD_res <- slim_run(AOD, throw_error = TRUE,parallel = TRUE)

```

### Extracting data and plotting

```{r}

df_rows <- ((gens-gen_bottleneck)/every_do) + 1
df_replicates_he <- as.data.frame(matrix(nrow =df_rows ,ncol = replicates ))
df_replicates_he_exp <- as.data.frame(matrix(nrow =df_rows ,ncol = replicates ))

for (i in 1:length(AOD_res)) {
  df_output <- AOD_res[[i]]
  df_output <- as.data.frame(cbind(df_output$output_data$generation,df_output$output_data$data,df_output$output_data$name))
  colnames(df_output) <- c("gen","He","name")
  df_output_het_neutral <- df_output[which(df_output$name=="het_neutral"),]
  df_small_pop <- df_output_het_neutral[which(df_output_het_neutral$gen==gen_bottleneck):nrow(df_output_het_neutral),]
  first_He <- as.numeric(df_output_het_neutral[which(df_output_het_neutral$gen==gen_bottleneck)-1,"He"])
  df_small_pop$rate_loss <- first_He * (rate_of_loss^(as.numeric(df_small_pop$gen)-gen_bottleneck) )
  df_replicates_he[,i] <- as.numeric(df_small_pop$He)
  df_replicates_he_exp[,i] <-  df_small_pop$rate_loss
}

df_replicates_he_res <- rowMeans(df_replicates_he)
df_replicates_he_exp_res <- rowMeans(df_replicates_he_exp)

plot_het <- as.data.frame(cbind(as.numeric(df_small_pop$gen), df_replicates_he_res,df_replicates_he_exp_res))
colnames(plot_het) <- c("gen", "He","he_exp")

all_sim <- print(ggplot(plot_het) +
                   geom_line(aes(x=gen,y=He),colour = "red",size=1) +
                   geom_line(aes(x=gen,y=he_exp),colour = "blue",size=1) +
                   theme_bw(base_size = 18) +
                   scale_fill_hc("darkunica")+
                   labs(x="GENERATIONS", y="He", title=NULL)+
                   theme(legend.title=element_blank())+
                   theme(legend.position =  "bottom") +
                   theme(legend.text=element_text(size=14)))

```

# Associative overdominance (AOD)

Overdominance arises when a heterozygote has a higher fitness than either homozygote. This is a type of balancing selection, where selection maintains two or more alleles in the population.

In the case of AOD, an apparent overdominance occurs if an individual has two different haplotypes and each haplotype contains recessive deleterious alleles ("a") at different positions (Figure below).

This arrangement of recessive deleterious alleles means that functionally normal alleles (alternative alleles "A") from one haplotype could mask the detrimental effects of deleterious alleles in the alternative haplotype. Hence, a homozygote of either haplotype will be less fit than a heterozygote, resulting in the maintenance of both haplotypes.

Eventually, this will break up due to recombination occurring within the haplotypes.

## **AOD's mechanism**

In this example for a diploid organism with a 3-locus heterozygous haplotype, deleterious recessive variants (red squares, "a") are masked by the alternative alleles (brown squares, "A") of the opposite haplotype. As a result, both haplotypes, alpha and beta, will be maintained by linked selection, along with alleles at a linked neutral locus (blue squares, "1" and "2"). However, recombination events (green arrows) can alter the linkage between loci which weaken the effects of AOD on genetic diversity.

![](AOD.jpg){width="730"}

## Visualize haplotypes

```{r opts_chunk$set(cache=F)}

# for the moment the following code only works for the last simulations that were performed

# minimum size of the haplotypes to look for
size_haplo <- 100
# number of loci of the window to calculate LD 
ld_depth <- 30

# which simulation to use 
sim <- AOD_res
# which replicate to use 
which_replicate <- 5
source('visual.R')

# number of deleterious mutations per individual
del_per_ind

```

```{r}
# haplotypes plot
p1
ggsave(paste0("haplo_1.pdf"),  width = 30, height =10, units = "in", dpi="retina", bg = "transparent" )

#to have a better look at the plot open the pdf file from the files pane
```

```{r echo=FALSE, opts_chunk$set(cache=F)}
# iterative plot without haplotypes (not supported by plotly yet)
p2

```
